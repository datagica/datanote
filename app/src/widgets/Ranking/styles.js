module.exports = {
  containerStyle: {
    display: 'inline-block',
    float: 'left',
    overflowX: 'hidden',
    padding: '0px 0px 0px 0px',
    margin: '0px 0px 0px 0px',
  },
  contentStyle: {
    margin: '5px 8%',
    padding: '10px 15px'
  },
  title: {
    /*
    OLD
    textShadow: 'white 0px 0px 1px',
    color: 'rgba(86, 96, 111, 0.7)',
    lineHeight: '18px',
    fontFamily: 'Roboto',
    letterSpacing: '0.4px',
    textTransform: 'uppercase',
    fontSize: '14px',
    fontWeight: '400',
    //textAlign: 'center',
    margin: '0px 0px 15px 0px',
    padding: '0px 0px 7px 0px',
    //borderBottom: 'solid 1px rgba(0, 0, 0, 0.07)',
    */
    textShadow: 'white 0px 0px 1px',
    color: 'rgba(89, 83, 93, 0.8)',
    lineHeight: '18px',
    fontFamily: 'Open Sans',
    letterSpacing: '0.2px',
    textTransform: 'uppercase',
    fontSize: '12.2px',
    fontWeight: 600,
    margin: '0px 0px 0px 0px',
    padding: '15px 0px 1px 26px',
  },
  item: {
    padding: '0px',
    margin: '0px',
    //display: 'block',
    display: 'flex',
    alignItems: 'center',
    //justifyContent: 'center'
    flexDirection: 'row'
  }
}
