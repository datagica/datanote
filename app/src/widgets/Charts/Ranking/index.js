'use strict'

import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Charts from '~/widgets/Charts'
import styles from '~/widgets/Charts/styles'

const Ranking = ({ label, data }) => (
  <Charts
    key={label.toLowerCase()}
    style={ styles.chart }
    size={ styles.chartSize }
    axis={{
      x: {
        type: 'category',
      },
      // For discrete values (enumerations) we rotate to have horizontal bars and
      // sorted list, since there can be many values.
      rotated: true
    }}
    bar={{
      width: {
        ratio: 0.8
      }
    }}
    data={{
      type: 'bar',
      json: data
        .sort((a, b) => b.value - a.value)
        .slice(0, 15), // no scroll yet so we have to crop the data
      keys: {
        x: 'x',
        value: [ 'value' ]
      },
      onclick: (d, element) => {
        console.log("Clicked on bar: ", { d: d, element: element })
      }
    }}
  />
)

export default Ranking
