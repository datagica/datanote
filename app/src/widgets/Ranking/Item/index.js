'use strict'

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styles from './styles.js'

import { Flex } from 'reflexbox'

class Item extends Component {

  static counter = 0;

  static propTypes = {
    label : PropTypes.object,
    rank  : PropTypes.number,
    ratio : PropTypes.number,
    height: PropTypes.number,
  };

  static get defaultProps(){
    return {
      rank  : 0,
      ratio : 0.5,
      height: 35
    }
  }

  render() {
    const label    = this.props.label;
    const height   = `${this.props.height}px`;
    const progress = `${this.props.ratio}%`;
    return (
      <Flex
        m={0} p={0}
        align='center'
        style={{ height: height }}>
        {(typeof label === 'undefined' || label === null) ? null :
          <div style={Object.assign({}, styles.itemLabel, { height: height, lineHeight: height })}>
          {(this.props.label && this.props.label.en) ? `${this.props.label.en}` : `${this.props.label}`}
        </div>}
        <div style={styles.itemProgressBg}>
          <div style={Object.assign({}, styles.itemProgressFg, { width: progress })}></div>
        </div>
        <div style={Object.assign({}, styles.itemProgressRank, { height: height, lineHeight: height })}>
          {this.props.rank}
        </div>
      </Flex>
    )
  }
}

export default Item
