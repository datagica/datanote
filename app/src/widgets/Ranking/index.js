'use strict'

import React, { Component } from 'react'
import PropTypes  from 'prop-types'
import boundaries from '~/utils/analysis/boundaries'

import Item from './Item'

import styles from './styles'

class Ranking extends Component {

  static counter = 0;

  static propTypes = {
    title:          PropTypes.string,
    data:           PropTypes.array,
    maxRank:        PropTypes.number,
    shownItems:     PropTypes.number,
    itemHeight:     PropTypes.number,
    fixHeight:      PropTypes.bool,
    scrollable:     PropTypes.bool,
    normalize:      PropTypes.bool,
    containerStyle: PropTypes.object,
    contentStyle:   PropTypes.object,
  }

  static get defaultProps(){
    return {
      title: "",
      data: [],
      maxRank: null,
      shownItems: 0,
      itemHeight: 35,
      fixHeight: false,
      scrollable: false,
      normalize: true,
      containerStyle: {},
      contentStyle: {}
    }
  }

  render() {

    const MAX_RANK = 1e6;
    let data = this.props.data;

    data.sort((a, b) => (b.rank - a.rank));

    // normally, we only show the first N items, and let the user scroll if
    // there is too many items. However, the user can also choose to disable
    // scroll. In that case, we optimize by cutting the rest of the array.
    if (this.props.shownItems && this.props.shownItems < data.length && !this.props.scrollable) {
      //console.log("slicing to "+this.shownItems)
      data = data.slice(0, this.shownItems)
    }

    let maxRank = !isNaN(this.props.maxRank) && isFinite(this.props.maxRank) ? this.props.maxRank : 100;
    //console.log("maxRank: ", maxRank)
    //console.log("normalize: ", this.props.normalize)
    if (this.props.normalize) {
      //console.log("DATA: "+JSON.stringify(data))
      maxRank = Math.min(Math.max(1, boundaries(data, item => item.rank).max), MAX_RANK);
      //console.log("boundaries: ", boundaries(data, item => item.rank))
    }

    const style = Object.assign({}, styles.ranking, {
      height: (this.props.fixHeight && this.props.shownItems)
        ? (this.props.itemHeight * this.props.shownItems + 10 + 25 )
        : 'auto',
      overflowY: this.props.scrollable ? 'auto' : 'hidden'
    })

    return (
      <div>
        <div style={Object.assign({}, styles.containerStyle, this.props.containerStyle)}>
          <h3 style={styles.title}>{this.props.title}</h3>
          <div style={Object.assign({}, styles.contentStyle, this.props.contentStyle)}>
          {data.map(item =>
            <Item
              key={item.id}
              label={item.label}
              rank={item.rank}
              ratio={(item.rank / maxRank) * 100}
              height={this.props.itemHeight}
            />
          )}
        </div>
      </div>
    </div>
     )
  }
}

export default Ranking;
