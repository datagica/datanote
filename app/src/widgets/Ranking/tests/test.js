import React    from 'react';
import renderer from 'react-test-renderer';

// mock the matchMedia (used by Flexbox)
window.matchMedia = window.matchMedia || function() {
    return {
        matches : false,
        addListener : function() {},
        removeListener: function() {}
    };
};

import Ranking  from '../index';


test('Default Ranking', () => {
  const component = renderer.create(<Ranking />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  //expect(sum(1, 2)).toBe(3);
});

test('Basic Ranking', () => {
  const component = renderer.create(
    <Ranking
      title={"Test"}
      data={[
        {
          id: "one",
          label: {
            en: "One",
          },
          rank: 50
        },
        {
          id: "two",
          label: {
            en: "Two"
          },
          rank: 25
        },
        {
          id: "three",
          label: {
            en: "Three"
          },
          rank: 100
        }
      ]}
      maxRank={100}
      shownItems={2}
      itemHeight={35}
      fixHeight={false}
      scrollable={false}
      normalize={true}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  //expect(sum(1, 2)).toBe(3);
});
