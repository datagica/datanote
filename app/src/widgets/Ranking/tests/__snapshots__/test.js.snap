// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Ranking 1`] = `
<div>
  <div
    style={
      Object {
        "display": "inline-block",
        "float": "left",
        "margin": "0px 0px 0px 0px",
        "overflowX": "hidden",
        "padding": "0px 0px 0px 0px",
      }
    }
  >
    <h3
      style={
        Object {
          "color": "rgba(89, 83, 93, 0.8)",
          "fontFamily": "Open Sans",
          "fontSize": "12.2px",
          "fontWeight": 600,
          "letterSpacing": "0.2px",
          "lineHeight": "18px",
          "margin": "0px 0px 0px 0px",
          "padding": "15px 0px 1px 26px",
          "textShadow": "white 0px 0px 1px",
          "textTransform": "uppercase",
        }
      }
    >
      Test
    </h3>
    <div
      style={
        Object {
          "margin": "5px 8%",
          "padding": "10px 15px",
        }
      }
    >
      <div
        className="Flex"
        style={
          Object {
            "WebkitAlignItems": "center",
            "alignItems": "center",
            "boxSizing": "border-box",
            "display": "flex",
            "height": "35px",
            "margin": 0,
            "msAlignItems": "center",
            "padding": 0,
          }
        }
      >
        <div
          style={
            Object {
              "alignSelf": "flex-start",
              "color": "rgba(34, 32, 62, 0.760784)",
              "fontFamily": "Open Sans",
              "fontSize": "11.6px",
              "fontWeight": "600",
              "height": "35px",
              "letterSpacing": "0.2px",
              "lineHeight": "35px",
              "overflow": "hidden",
              "padding": "0px",
              "textOverflow": "ellipsis",
              "textShadow": "0px 0px 1px #ffffff",
              "textTransform": "capitalize",
              "whiteSpace": "nowrap",
              "width": "160px",
            }
          }
        >
          Three
        </div>
        <div
          style={
            Object {
              "background": "linear-gradient(rgba(255, 255, 255, 0.32) 0%, rgba(234, 234, 234, 0.12) 9%, rgba(218, 218, 218, 0.34) 100%)",
              "borderRadius": "5px",
              "boxShadow": "rgba(0, 0, 0, 0.26) 0px 1px 3px -1px inset",
              "flexGrow": "1",
              "height": "12px",
              "margin": "0",
              "overflow": "hidden",
              "padding": "0",
              "width": "50px",
            }
          }
        >
          <div
            style={
              Object {
                "background": "linear-gradient(rgb(192, 194, 208) 0%, rgb(149, 138, 208) 100%)",
                "borderRadius": "2px",
                "boxShadow": "rgba(0, 0, 0, 0.24) 0px 1px 0px 0px",
                "height": "13px",
                "margin": "0",
                "padding": "0",
                "width": "100%",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "color": "rgba(34, 32, 62, 0.760784)",
              "fontFamily": "Open Sans",
              "fontSize": "11.6px",
              "fontWeight": "600",
              "height": "35px",
              "letterSpacing": "0.2px",
              "lineHeight": "35px",
              "margin": "-1px 0px 0px 0px",
              "padding": "0px 0px 0px 10px",
              "textShadow": "0px 0px 1px #ffffff",
              "textTransform": "capitalize",
              "width": "20px",
            }
          }
        >
          100
        </div>
      </div>
      <div
        className="Flex"
        style={
          Object {
            "WebkitAlignItems": "center",
            "alignItems": "center",
            "boxSizing": "border-box",
            "display": "flex",
            "height": "35px",
            "margin": 0,
            "msAlignItems": "center",
            "padding": 0,
          }
        }
      >
        <div
          style={
            Object {
              "alignSelf": "flex-start",
              "color": "rgba(34, 32, 62, 0.760784)",
              "fontFamily": "Open Sans",
              "fontSize": "11.6px",
              "fontWeight": "600",
              "height": "35px",
              "letterSpacing": "0.2px",
              "lineHeight": "35px",
              "overflow": "hidden",
              "padding": "0px",
              "textOverflow": "ellipsis",
              "textShadow": "0px 0px 1px #ffffff",
              "textTransform": "capitalize",
              "whiteSpace": "nowrap",
              "width": "160px",
            }
          }
        >
          One
        </div>
        <div
          style={
            Object {
              "background": "linear-gradient(rgba(255, 255, 255, 0.32) 0%, rgba(234, 234, 234, 0.12) 9%, rgba(218, 218, 218, 0.34) 100%)",
              "borderRadius": "5px",
              "boxShadow": "rgba(0, 0, 0, 0.26) 0px 1px 3px -1px inset",
              "flexGrow": "1",
              "height": "12px",
              "margin": "0",
              "overflow": "hidden",
              "padding": "0",
              "width": "50px",
            }
          }
        >
          <div
            style={
              Object {
                "background": "linear-gradient(rgb(192, 194, 208) 0%, rgb(149, 138, 208) 100%)",
                "borderRadius": "2px",
                "boxShadow": "rgba(0, 0, 0, 0.24) 0px 1px 0px 0px",
                "height": "13px",
                "margin": "0",
                "padding": "0",
                "width": "50%",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "color": "rgba(34, 32, 62, 0.760784)",
              "fontFamily": "Open Sans",
              "fontSize": "11.6px",
              "fontWeight": "600",
              "height": "35px",
              "letterSpacing": "0.2px",
              "lineHeight": "35px",
              "margin": "-1px 0px 0px 0px",
              "padding": "0px 0px 0px 10px",
              "textShadow": "0px 0px 1px #ffffff",
              "textTransform": "capitalize",
              "width": "20px",
            }
          }
        >
          50
        </div>
      </div>
      <div
        className="Flex"
        style={
          Object {
            "WebkitAlignItems": "center",
            "alignItems": "center",
            "boxSizing": "border-box",
            "display": "flex",
            "height": "35px",
            "margin": 0,
            "msAlignItems": "center",
            "padding": 0,
          }
        }
      >
        <div
          style={
            Object {
              "alignSelf": "flex-start",
              "color": "rgba(34, 32, 62, 0.760784)",
              "fontFamily": "Open Sans",
              "fontSize": "11.6px",
              "fontWeight": "600",
              "height": "35px",
              "letterSpacing": "0.2px",
              "lineHeight": "35px",
              "overflow": "hidden",
              "padding": "0px",
              "textOverflow": "ellipsis",
              "textShadow": "0px 0px 1px #ffffff",
              "textTransform": "capitalize",
              "whiteSpace": "nowrap",
              "width": "160px",
            }
          }
        >
          Two
        </div>
        <div
          style={
            Object {
              "background": "linear-gradient(rgba(255, 255, 255, 0.32) 0%, rgba(234, 234, 234, 0.12) 9%, rgba(218, 218, 218, 0.34) 100%)",
              "borderRadius": "5px",
              "boxShadow": "rgba(0, 0, 0, 0.26) 0px 1px 3px -1px inset",
              "flexGrow": "1",
              "height": "12px",
              "margin": "0",
              "overflow": "hidden",
              "padding": "0",
              "width": "50px",
            }
          }
        >
          <div
            style={
              Object {
                "background": "linear-gradient(rgb(192, 194, 208) 0%, rgb(149, 138, 208) 100%)",
                "borderRadius": "2px",
                "boxShadow": "rgba(0, 0, 0, 0.24) 0px 1px 0px 0px",
                "height": "13px",
                "margin": "0",
                "padding": "0",
                "width": "25%",
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "color": "rgba(34, 32, 62, 0.760784)",
              "fontFamily": "Open Sans",
              "fontSize": "11.6px",
              "fontWeight": "600",
              "height": "35px",
              "letterSpacing": "0.2px",
              "lineHeight": "35px",
              "margin": "-1px 0px 0px 0px",
              "padding": "0px 0px 0px 10px",
              "textShadow": "0px 0px 1px #ffffff",
              "textTransform": "capitalize",
              "width": "20px",
            }
          }
        >
          25
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Default Ranking 1`] = `
<div>
  <div
    style={
      Object {
        "display": "inline-block",
        "float": "left",
        "margin": "0px 0px 0px 0px",
        "overflowX": "hidden",
        "padding": "0px 0px 0px 0px",
      }
    }
  >
    <h3
      style={
        Object {
          "color": "rgba(89, 83, 93, 0.8)",
          "fontFamily": "Open Sans",
          "fontSize": "12.2px",
          "fontWeight": 600,
          "letterSpacing": "0.2px",
          "lineHeight": "18px",
          "margin": "0px 0px 0px 0px",
          "padding": "15px 0px 1px 26px",
          "textShadow": "white 0px 0px 1px",
          "textTransform": "uppercase",
        }
      }
    >
      
    </h3>
    <div
      style={
        Object {
          "margin": "5px 8%",
          "padding": "10px 15px",
        }
      }
    />
  </div>
</div>
`;
